massifs/mmriver.go

```diff
@@ -0,0 +1,246 @@
+// Note that MMRIVER receipts allow for multiple inclusion proofs to be attached to the receipt.
+// This function returns true only if ALL receipts verify
+func VerifySignedInclusionReceipts(
+	ctx context.Context,
+	receipt *commoncose.CoseSign1Message,
+	candidates [][]byte,
+) (bool, []byte, error) {
+
+	var err error
+
+	// ignore any existing payload
+	receipt.Payload = nil
+
+	// We must return false if there are no candidates
+	if len(candidates) == 0 {
+		return false, nil, fmt.Errorf("no candidates provided")
+	}
+
+	var header MMRiverVerifiableProofsHeader
+	err = cbor.Unmarshal(receipt.Headers.RawUnprotected, &header)
+	if err != nil {
+		return false, nil, fmt.Errorf("MMRIVER receipt proofs malformed")
+	}
+	verifiableProofs := header.VerifiableProofs
+	if len(verifiableProofs.InclusionProofs) == 0 {
+		return false, nil, fmt.Errorf("MMRIVER receipt inclusion proofs not present")
+	}
+
+	// permit *fewer* candidates than proofs, but not more
+	if len(candidates) > len(verifiableProofs.InclusionProofs) {
+		return false, nil, fmt.Errorf("MMRIVER receipt more candidates than proofs")
+	}
+
+	var proof MMRiverInclusionProof
+
+	proof = verifiableProofs.InclusionProofs[0]
+	receipt.Payload = mmr.IncludedRoot(
+		sha256.New(),
+		proof.Index, candidates[0],
+		proof.InclusionPath)
+
+	err = receipt.VerifyWithCWTPublicKey(nil)
+	if err != nil {
+		return false, nil, fmt.Errorf(
+			"MMRIVER receipt VERIFY FAILED for: mmrIndex %d, candidate %d, err %v", proof.Index, 0, err)
+	}
+	// verify the first proof then just compare the produced roots
+
+	for i := 1; i < len(verifiableProofs.InclusionProofs); i++ {
+
+		proof = verifiableProofs.InclusionProofs[i]
+		proven := mmr.IncludedRoot(sha256.New(), proof.Index, candidates[i], proof.InclusionPath)
+		if bytes.Compare(receipt.Payload, proven) != 0 {
+			return false, nil, fmt.Errorf(
+				"MMRIVER receipt VERIFY FAILED for: mmrIndex %d, candidate %d, err %v", proof.Index, i, err)
+		}
+	}
+	return true, receipt.Payload, nil
+}
+
+// VerifySignedInclusionReceipt verifies a reciept comprised of a single inclusion proof
+// If there are 0 or more than 1 candidates, the result will be false and an error will be returned
+func VerifySignedInclusionReceipt(
+	ctx context.Context,
+	receipt *commoncose.CoseSign1Message,
+	candidate []byte,
+) (bool, []byte, error) {
+
+	ok, root, err := VerifySignedInclusionReceipts(ctx, receipt, [][]byte{candidate})
+	if err != nil {
+		return false, nil, err
+	}
+	if !ok {
+		return false, nil, nil
+	}
+	return true, root, nil
+}
+
+type verifiedContextGetter interface {
+	GetVerifiedContext(
+		ctx context.Context, tenantIdentity string, massifIndex uint64,
+		opts ...ReaderOption,
+	) (*VerifiedContext, error)
+}
+
+// GetReceipt returns a COSE receipt for the given tenantIdentity and mmrIndex
+func NewReceipt(
+	ctx context.Context,
+	massifHeight uint8,
+	tenantIdentity string, mmrIndex uint64,
```

Comment:
minor: should we start using the term logTenant?

Reply:
yes. definitely. I might do that as a follow up in the 'redux' pr as there will be some references to be careful of